modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java

```diff
@@ -60,24 +65,46 @@
     private static final ObjectParser<ReindexRequest, ReindexParseContext> PARSER = new ObjectParser<>("reindex");
 
     static {
-        ObjectParser.Parser<SearchRequest, ReindexParseContext> sourceParser = (parser, search, context) -> {
-            /*
-             * Extract the parameters that we need from the source sent to the parser. We could do away with this hack when search source
-             * has an ObjectParser.
-             */
+        ObjectParser.Parser<ReindexRequest, ReindexParseContext> sourceParser = (parser, request, context) -> {
+            // Funky hack to work around Search not having a proper ObjectParser and us wanting to extract query if using remote.
             Map<String, Object> source = parser.map();
             String[] indices = extractStringArray(source, "index");
             if (indices != null) {
-                search.indices(indices);
+                request.getSearchRequest().indices(indices);
             }
             String[] types = extractStringArray(source, "type");
             if (types != null) {
-                search.types(types);
+                request.getSearchRequest().types(types);
+            }
+            @SuppressWarnings("unchecked")
+            Map<String, Object> remote = (Map<String, Object>) source.remove("remote");
+            if (remote != null) {
+                String hostInRequest = requireNonNull(extractString(remote, "host"),
+                        "[host] must be specified to reindex from a remote cluster");
+                String[] split = hostInRequest.split(":");
```

Comment:
This is totally going to bork things later on unless we split schema apart. I'll leave a larger comment about this when I'm done (this is a reminder to me)